### PR TITLE
[Merged by Bors] - Deprecate `.system`

### PR DESCRIPTION
**Pull request by DJMcNab**
# Objective

- Using `.system()` is no longer idiomatic.

## Solution

- Give a warning when using it

### CODE REVIEW

File: crates/bevy_ecs/src/system/function_system.rs [hunk 1]
```diff
@@ -249,24 +249,42 @@ impl<P: SystemParam + 'static> System for ParamSystem<P> {
...
+    /// However, since [#2398](https://github.com/bevyengine/bevy/pull/2398),
+    /// this is no longer required.
+    ///
+    /// In future, this method will be removed.
+    ///
+    /// One valid use of this method is to assert that a given function is a valid system.
+    /// For this case, use `IntoSystem::into_system` instead.
+    #[deprecated(
+        since = "0.6.0",
+        note = "In most cases, using `.system()` is no longer necessary, and so it can be removed"
```
Comment (NiklasEi) [inline crates/bevy_ecs/src/system/function_system.rs:272]: This reads as if there are still cases that require `.system()`. Is that the case? If so, what happens with them when we remove it (like announced in the docs)? Should we actually deprecate it then?

File: crates/bevy_ecs/src/system/function_system.rs [hunk 2]
```diff
@@ -249,24 +249,42 @@ impl<P: SystemParam + 'static> System for ParamSystem<P> {
...
     type System: System<In = In, Out = Out>;
     /// Turns this value into its corresponding [`System`].
-    fn system(self) -> Self::System;
+    ///
+    /// Use of this method was formerly required whenever adding a `system` to an `App`.
+    /// or other cases where a system is required.
+    /// However, since [#2398](https://github.com/bevyengine/bevy/pull/2398),
+    /// this is no longer required.
+    ///
+    /// In future, this method will be removed.
```
Comment (NiklasEi) [inline crates/bevy_ecs/src/system/function_system.rs:266]:
```suggestion
    /// In the future, this method will be removed.
```

**Comment by cart** [comment]
I think my biggest problem with this is that it isn't _really_ deprecated in all contexts. If a user is building some custom collection of systems and wants to store them in a `Vec<Box<dyn System>>`, they need to call `Box::new(some_system.system())`. No way around it!

Given that we have no plans remove `.system()` entirely, as it is an integral part of system implementation, there is no real harm in letting people call `.system()` where they want to. People won't want to generally because it is just extra boilerplate and that will probably be enough incentive for existing code to migrate. And if we remove `.system()` from our examples and docs, new users won't even need to know it exists.

**Comment by inodentry** [comment]
I agree. Marking it as deprecated seems unnecessary. We aren't removing it, and I doubt users would want to stubbornly continue using it anyway.

File: crates/bevy_ecs/src/system/function_system.rs (hunk 1 above)
Comment (DJMcNab) [reply crates/bevy_ecs/src/system/function_system.rs:272]: This is the case explained directly in the line above, which is testing that a function is a system.

**Comment by DJMcNab** [comment]
Perhaps?

I'd claim that managing a `Vec<Box<dyn System>>` is not something we should be wanting users to be doing, given how error prone it is (adding new archetypes needs to be manually handled, they don't get the parallelism benefits, etc.). I don't know of any manual scheduler implementations outside of our crates, which is the single case that you're arguing for the ergonomics of, at the expense of "occasionally this weird syntax pops up".

Also, there is a way around it, which is to use `IntoSystem::into_system` to convert them into a system explicitly. My claim is that having the method around as a method instead of as an (associated) free function is not correct, given how rarely it is correctly used.

In my view, the only context in which an end user could correctly be using `.system()` currently is when creating a test. I think this would be better served by an `assert_is_system(it)`, which probably should be part of this pr.

I'm only deprecating it in this pr rather than removing it entirely because it used to be required in `0.5`, and so it's a nicer migration path if it continues to work for one release.

**Comment by mockersf** [comment]
> In my view, the only context in which an end user could correctly be using `.system()` currently is when creating a test. I think this would be better served by an `assert_is_system(it)`, which probably should be part of this pr.

I love that part. 
I like the rest of the PR, but I love the assert method instead of just calling `.system()` in the doc comment

File: crates/bevy_ecs/src/system/function_system.rs [hunk 3]
```diff
@@ -249,24 +249,42 @@ impl<P: SystemParam + 'static> System for ParamSystem<P> {
...
+    ///
+    /// One valid use of this method is to assert that a given function is a valid system.
+    /// For this case, use `IntoSystem::into_system` instead.
```
Comment (IceSentry) [inline crates/bevy_ecs/src/system/function_system.rs:269]:
```suggestion
    /// For this case, use `bevy_ecs::system::assert_is_system` instead.
```

**Review by IceSentry — COMMENTED** [review]
I think the removal of `.system()` internally and adding the deprecation warning should be separate PRs. Whether or not `.system()` is marked as deprecated, removing the use of it internally is still a good thing in my opinion.

**Comment by alice-i-cecile** [comment]
> I don't know of any manual scheduler implementations outside of our crates, which is the single case that you're arguing for the ergonomics of, at the expense of "occasionally this weird syntax pops up".

Agreed. As one of the folks most likely to *write* an external manual scheduler, I'm perfectly okay with slightly sacrificing the ergonomics there in order to provide a nice deprecation notice to our casual users and to reduce our API surface. If I really care while writing fancy custom schedulers, I'll write my own extension method to get it back.

For the most part though, writing external schedulers should be vanishingly rare: the initialization and archetype logic is too risky to be touching without an extremely good reason. We should be working towards making the general API flexible enough, with good presets and/or examples for common use cases (like turn-based games).

If we can get the `assert_is_system` function into this PR, I'm on board.

**Comment by DJMcNab** [comment]
`assert_is_system` should already be in this PR?

**Review by alice-i-cecile — APPROVED** [review]

**Comment by cart** [comment]
Ok I've come around on this. `.system()` _really_ isn't needed anymore (official bevy Schedules and 3rd party schedules should both favor abstracted IntoSystem calls). My hard headedness stemmed from missing the new `into_system` call (which functionally replaces `.system()`, but just removes the ability to call it like a "method"). Sorry for not wrapping my head around this sooner.

**Pull request closed without merging**